File: go-server/protocol/protocol.go
package protocol

import (
	"encoding/binary"
	"errors"
	"math"
	"regexp"
)

var validUsername = regexp.MustCompile(`^[a-zA-Z0-9_-]{1,16}$`)

type EntityState struct {
	ID         uint16
	Type       uint8
	Subtype    uint8
	X          float32
	Y          float32
	Angle      float32
	Health     uint16
	MaxHealth  uint16
	Radius     uint16
	StateFlags uint32
}

type ClientInput struct {
	Keys             uint8
	MouseAngle       float32
	UpgradeSelect    uint8
	DeleteSlotSelect uint8
}

func EncodeWelcome(playerID uint16, arenaWidth, arenaHeight float32) []byte {
	buf := make([]byte, 11)
	buf[0] = 1
	binary.LittleEndian.PutUint16(buf[1:3], playerID)
	binary.LittleEndian.PutUint32(buf[3:7], math.Float32bits(arenaWidth))
	binary.LittleEndian.PutUint32(buf[7:11], math.Float32bits(arenaHeight))
	return buf
}

func EncodeWorldState(tick uint32, xp, maxXP uint32, level uint16, score uint32, health, maxHealth uint16, upgradePoints uint8, upgradeLevels []uint8, waveNumber uint16, card1, card2, card3 uint8, inventory []uint8, entities []EntityState, removedIDs []uint16) []byte {
	count := len(entities)
	removedCount := len(removedIDs)
	bufSize := 258 + count*26 + removedCount*2
	buf := make([]byte, bufSize)
	buf[0] = 2
	binary.LittleEndian.PutUint32(buf[1:5], tick)
	binary.LittleEndian.PutUint32(buf[5:9], xp)
	binary.LittleEndian.PutUint32(buf[9:13], maxXP)
	binary.LittleEndian.PutUint16(buf[13:15], level)
	binary.LittleEndian.PutUint32(buf[15:19], score)
	binary.LittleEndian.PutUint16(buf[19:21], health)
	binary.LittleEndian.PutUint16(buf[21:23], maxHealth)
	buf[23] = upgradePoints
	for i := 0; i < 25; i++ {
		if i < len(upgradeLevels) {
			buf[24+i] = upgradeLevels[i]
		} else {
			buf[24+i] = 0
		}
	}
	binary.LittleEndian.PutUint16(buf[49:51], waveNumber)
	binary.LittleEndian.PutUint16(buf[51:53], uint16(count))
	buf[53] = card1
	buf[54] = card2
	buf[55] = card3
	for i := 0; i < 200; i++ {
		if i < len(inventory) {
			buf[56+i] = inventory[i]
		} else {
			buf[56+i] = 0
		}
	}
	binary.LittleEndian.PutUint16(buf[256:258], uint16(removedCount))

	offset := 258
	for i := 0; i < count; i++ {
		ent := &entities[i]
		binary.LittleEndian.PutUint16(buf[offset:offset+2], ent.ID)
		buf[offset+2] = ent.Type
		buf[offset+3] = ent.Subtype
		binary.LittleEndian.PutUint32(buf[offset+4:offset+8], math.Float32bits(ent.X))
		binary.LittleEndian.PutUint32(buf[offset+8:offset+12], math.Float32bits(ent.Y))
		binary.LittleEndian.PutUint32(buf[offset+12:offset+16], math.Float32bits(ent.Angle))
		binary.LittleEndian.PutUint16(buf[offset+16:offset+18], ent.Health)
		binary.LittleEndian.PutUint16(buf[offset+18:offset+20], ent.MaxHealth)
		binary.LittleEndian.PutUint16(buf[offset+20:offset+22], ent.Radius)
		binary.LittleEndian.PutUint32(buf[offset+22:offset+26], ent.StateFlags)
		offset += 26
	}

	for i := 0; i < removedCount; i++ {
		binary.LittleEndian.PutUint16(buf[offset:offset+2], removedIDs[i])
		offset += 2
	}

	return buf
}

func EncodeGameOver(score uint32, wave uint32) []byte {
	buf := make([]byte, 9)
	buf[0] = 4
	binary.LittleEndian.PutUint32(buf[1:5], score)
	binary.LittleEndian.PutUint32(buf[5:9], wave)
	return buf
}

func DecodeJoin(buf []byte) (string, uint8, error) {
	if len(buf) < 2 {
		return "", 0, errors.New("invalid join size")
	}
	usernameLen := int(buf[1])
	if usernameLen == 0 || usernameLen > 16 {
		return "", 0, errors.New("invalid username length")
	}
	if len(buf) < 3+usernameLen {
		return "", 0, errors.New("invalid join size for username and class")
	}
	username := string(buf[2 : 2+usernameLen])
	if !validUsername.MatchString(username) {
		return "", 0, errors.New("invalid characters in username")
	}
	classId := buf[2+usernameLen]
	return username, classId, nil
}

func DecodeInput(buf []byte) (ClientInput, error) {
	if len(buf) < 8 {
		return ClientInput{}, errors.New("invalid input size")
	}
	var input ClientInput
	input.Keys = buf[1]
	input.MouseAngle = math.Float32frombits(binary.LittleEndian.Uint32(buf[2:6]))
	input.UpgradeSelect = buf[6]
	input.DeleteSlotSelect = buf[7]
	return input, nil
}

func DecodeUpgradeClass(buf []byte) (uint8, error) {
	if len(buf) < 2 {
		return 0, errors.New("invalid upgrade class size")
	}
	return buf[1], nil
}

File: ts-client/src/protocol.ts
export interface WelcomeMessage {
  type: "welcome";
  playerId: number;
  arenaWidth: number;
  arenaHeight: number;
}

export interface EntityState {
  id: number;
  type: number;
  subtype: number;
  x: number;
  y: number;
  angle: number;
  health: number;
  maxHealth: number;
  radius: number;
  stateFlags: number;
}

export interface WorldStateMessage {
  type: "worldState";
  tick: number;
  xp: number;
  maxXp: number;
  level: number;
  score: number;
  health: number;
  maxHealth: number;
  upgradePoints: number;
  upgrades: number[];
  waveNumber: number;
  card1: number;
  card2: number;
  card3: number;
  inventory: number[];
  entities: EntityState[];
  removedIds: number[];
}

export interface GameOverMessage {
  type: "gameOver";
  score: number;
  wave: number;
}

export type GameMessage = WelcomeMessage | WorldStateMessage | GameOverMessage;

export function serializeJoin(username: string, classId: number): ArrayBuffer {
  const encoder = new TextEncoder();
  const nameBytes = encoder.encode(username);
  const buffer = new ArrayBuffer(3 + nameBytes.length);
  const view = new DataView(buffer);
  view.setUint8(0, 1);
  view.setUint8(1, nameBytes.length);
  const u8Array = new Uint8Array(buffer, 2);
  u8Array.set(nameBytes);
  view.setUint8(2 + nameBytes.length, classId);
  return buffer;
}

export function serializeInput(keys: number, mouseAngle: number, upgradeSelect: number, deleteSlotSelect: number): ArrayBuffer {
  const buffer = new ArrayBuffer(8);
  const view = new DataView(buffer);
  view.setUint8(0, 2);
  view.setUint8(1, keys);
  view.setFloat32(2, mouseAngle, true);
  view.setUint8(6, upgradeSelect);
  view.setUint8(7, deleteSlotSelect);
  return buffer;
}

export function deserializeMessage(buffer: ArrayBuffer): GameMessage | null {
  const view = new DataView(buffer);
  if (view.byteLength < 1) return null;
  const opcode = view.getUint8(0);
  if (opcode === 1) {
    if (view.byteLength < 11) return null;
    return {
      type: "welcome",
      playerId: view.getUint16(1, true),
      arenaWidth: view.getFloat32(3, true),
      arenaHeight: view.getFloat32(7, true),
    };
  } else if (opcode === 2) {
    if (view.byteLength < 262) return null;
    const tick = view.getUint32(1, true);
    const xp = view.getUint32(5, true);
    const maxXp = view.getUint32(9, true);
    const level = view.getUint16(13, true);
    const score = view.getUint32(15, true);
    const health = view.getUint16(19, true);
    const maxHealth = view.getUint16(21, true);
    const upgradePoints = view.getUint8(23);
    const upgrades: number[] = [];
    for (let i = 0; i < 29; i++) {
      upgrades.push(view.getUint8(24 + i));
    }
    const waveNumber = view.getUint16(53, true);
    const entitiesCount = view.getUint16(55, true);
    const card1 = view.getUint8(57);
    const card2 = view.getUint8(58);
    const card3 = view.getUint8(59);
    const inventory: number[] = [];
    for (let i = 0; i < 200; i++) {
      inventory.push(view.getUint8(60 + i));
    }
    const removedCount = view.getUint16(260, true);

    const entities: EntityState[] = [];
    let offset = 262;
    for (let i = 0; i < entitiesCount; i++) {
      if (offset + 26 > view.byteLength) break;
      entities.push({
        id: view.getUint16(offset, true),
        type: view.getUint8(offset + 2),
        subtype: view.getUint8(offset + 3),
        x: view.getFloat32(offset + 4, true),
        y: view.getFloat32(offset + 8, true),
        angle: view.getFloat32(offset + 12, true),
        health: view.getUint16(offset + 16, true),
        maxHealth: view.getUint16(offset + 18, true),
        radius: view.getUint16(offset + 20, true),
        stateFlags: view.getUint32(offset + 22, true),
      });
      offset += 26;
    }

    const removedIds: number[] = [];
    for (let i = 0; i < removedCount; i++) {
      if (offset + 2 > view.byteLength) break;
      removedIds.push(view.getUint16(offset, true));
      offset += 2;
    }

    return {
      type: "worldState",
      tick,
      xp,
      maxXp,
      level,
      score,
      health,
      maxHealth,
      upgradePoints,
      upgrades,
      waveNumber,
      card1,
      card2,
      card3,
      inventory,
      entities,
      removedIds,
    };
  } else if (opcode === 4) {
    if (view.byteLength < 9) return null;
    return {
      type: "gameOver",
      score: view.getUint32(1, true),
      wave: view.getUint32(5, true),
    };
  }
  return null;
}
